tests/monitoring_processes/test_monitoring_critical_processes.py

```diff
@@ -0,0 +1,543 @@
+
+    Args:
+        duthost: Hostname of DUT.
+        container_name: A string shows name of a container.
+        critical_process: A string shows name of a process.
+
+    Returns:
+        None.
+    """
+    logger.info("Checking whether process '{}' in container '{}' is running...".format(critical_process, container_name))
+    program_status, program_pid = get_program_info(duthost, container_name, critical_process)
+    if program_status == "RUNNING":
+        logger.info("Process '{}' in container '{} is running.".format(critical_process, container_name))
+    else:
+        logger.info("Process '{}' in container '{}' is not running and restart it...".format(critical_process, container_name))
+        command_output = duthost.shell("docker exec {} supervisorctl restart {}".format(container_name, critical_process))
+        if command_output["rc"] == 0:
+            logger.info("Process '{}' in container '{}' is restarted.".format(critical_process, container_name))
+        else:
+            pytest.fail("Failed to restart process '{}' in container '{}'.".format(critical_process, container_name))
+
+
+def restart_critical_processes(duthost, containers_in_namespaces):
+    """Restarts all critical process in enabled containers.
+
+    Args:
+        duthost: Hostname of DUT.
+        containers_in_namespaces: A dictionary where keys are container names and
+        values are lists which contains ids of namespaces this container should reside in.
+
+    Returns:
+        None.
+    """
+    for container_name in containers_in_namespaces.keys():
+        critical_group_list, critical_process_list, succeeded = duthost.get_critical_group_and_process_lists(container_name)
+        pytest_assert(succeeded, "Failed to get critical group and process lists of container '{}'".format(container_name))
+
+        namespaces = containers_in_namespaces[container_name]
+        for namespace_id in namespaces:
+            container_name_in_namespace = container_name
+            if namespace_id != "host":
+                container_name_in_namespace += namesapce_id
+
+            for critical_process in critical_process_list:
+                # Skip 'dsserve' process since it was not managed by supervisord
+                # TODO: Should remove the following two lines once the issue was solved in the image.
+                if container_name_in_namespace == "syncd" and critical_process == "dsserve":
+                    continue
+
+                check_and_restart_process(duthost, container_name_in_namespace, critical_process)
+
+            for critical_group in critical_group_list:
+                group_program_info = get_group_program_info(duthost, container_name_in_namespace, critical_group)
+                for program_name in group_program_info:
+                    check_and_restart_process(duthost, container_name_in_namespace, program_name)
+
+
+def restore_containers_autorestart(duthost, containers_autorestart_states):
```

Comment:
You should combine this function with disable autostart function to be a fixture.

You should only reenable autorestart for contains you disabled earlier.

Reply:
Added a fixture to disabled/enable the autorestart before/after the testing.